CL-1881 Model project status with a fact table

adessy opened the pull request.

adessy committed:
CL-1881 Implement view for project statuses (fact table)

adessy committed:
CL-1881 Rework fact_project_statuses view

A `date` column is not sufficient to define a global order on the status changes, so I added a `timestamp` column.

jamesspeake commented on back/engines/commercial/analytics/db/views/analytics_fact_project_statuses_v01.sql:
```diff
@@ -0,0 +1,37 @@
+         SELECT * FROM finished_statuses_for_continuous_projects
+     )
+
+SELECT *, timestamp::DATE as date
```
I would be specific here about which columns as when I run it there is a timestamp column that I don't think we need

adessy replied:
As discussed, keeping it for now as it is needed to get a global ordering when there were multiple status change the same day.

jamesspeake commented on back/engines/commercial/analytics/db/views/analytics_fact_project_statuses_v01.sql:
```diff
@@ -0,0 +1,37 @@
+WITH regular_project_statuses AS -- project statuses that do not need to be calculated
+         (SELECT item_id as project_id, action as status, acted_at as timestamp
```
we've now got a bit of a naming convention of using dimension_ for all the dimension columns

jamesspeake replied:
so in the output view here we'd have dimension_project_id and dimension_data_id - id is needed on the date column so that rails can use it as a key

jamesspeake commented on back/engines/commercial/analytics/db/views/analytics_fact_project_statuses_v01.sql:
```diff
@@ -0,0 +1,37 @@
+
+     last_statuses_for_continuous_projects AS
+         (SELECT DISTINCT ON (project_id) *
+          FROM regular_project_statuses JOIN projects ON project_id = projects.id
```
point of style really, but I would always favour using aliases when joining tables

jamesspeake reviewed (commented):
I've had a play around with the query and this looks like it should be able to summarise everything we want. A few comments for you in there, but nothing that changes the logic.

I've not really used the WITH syntax much in the past but this is pretty neat for dividing it all up.

I do also wonder if something like this is a case for starting to use materialised views. It could be searching through a lot of data in the activities table to get these figures, and there are sub queries in there, so could be worth testing the SQL against one of our larger tenants. I've been meaning to do this for a few of our views TBH.

adessy committed:
CL-1881 Update Gemfile.lock

adessy committed:
Merge branch 'master' into CL-1881/fact-project-status

adessy committed:
CL-1881 Create fact_project_statuses view

adessy committed:
CL-1881 Add FactProjectStatus model

adessy committed:
CL-1881 Allow to query project_status fact table

Commit:
Translations updated by CI (extract-intl)

adessy committed:
CL-1881 Rework the json-schema definition for analytics queries

The fact enum in the json-schema is no longer hardcoded. It gets its value from Analytics::Query.fact_names.

The json-schema is memoized at the class level.

adessy committed:
Merge remote-tracking branch 'origin/CL-1881/fact-project-status' into CL-1881/fact-project-status

adessy committed:
CL-1881 Minor refactoring

adessy committed:
CL-1881 Fix failing test

adessy committed:
CL-1881 Rework Analytics::Query#fact_attributes

- Return early if attributes have already been calculated
- Don't rely on options in the association definition to infer the association class. This means that `fact_attributes` now also works for associations where the class_name has not been specified explicitly.

adessy committed:
CL-1881 Rework the analytics_fact_project_statuses view

Rework the view to make it easier for us down the road.

The new view only lists the last status. The main reason for this change is that, in its current state, the analytics API does not allow us to query the last status of each project (it has a minimal 'first' aggregation method but data in each group cannot be sorted).

Another small change is that *draft* timeline projects whose last phase is complete are no longer considered complete.

We also dropped the 'created' and 'deleted' statuses. The 'created' was pretty much useless anyway (it could be inferred from other status changes). The 'deleted' status is kind of interesting to know, but including it removes deleted projects from the list of finished projects (since we only consider the last status). Since we don't show any information about deleted projects in UI, I decided to drop it for now.

adessy committed:
CL-1881 Update schema.rb

See 0f3c797948fdd3f7a265bc88dde4f0351653084b

adessy committed:
CL-1881 Rework the analytics_fact_project_statuses view

Contrary to what was said in 0f3c797948fdd3f7a265bc88dde4f0351653084b, we must include the deleted status, otherwise deleted projects would still count as published, draft, archived based on their last status (before deletion).

As a result, deleting a project also removes its finished status, if any.

adessy committed:
CL-1881 Add acceptance tests for project status analytics

adessy committed:
CL-1881 Fix Analytics::Query#fact_attributes

It was returning nil when @fact_attributes were already computed.

adessy committed:
CL-1881 Adapt tests in fact_project_status_spec.rb

Adapt fact_project_status_spec.rb to the new definition of the analytics_fact_project_statuses view. See: 0f3c797948fdd3f7a265bc88dde4f0351653084b.

adessy committed:
CL-1881 Fix rubocop offense

adessy committed:
CL-1881 Add some documentation [skip ci]

adessy committed:
Merge branch 'master' into CL-1881/fact-project-status

jamesspeake commented on back/engines/commercial/analytics/app/models/analytics/fact_project_status.rb:
```diff
@@ -0,0 +1,17 @@
+module Analytics
+  class FactProjectStatus < Analytics::ApplicationRecordView
+    belongs_to :dimension_date, foreign_key: :date
+    belongs_to :project
```
Because this in the analytics module This needs to belong to dimension_project not project, although they're not much different, they are different types

jamesspeake commented on back/engines/commercial/analytics/app/services/analytics/query_validator_service.rb:
```diff
@@ -2,9 +2,13 @@
 
 module Analytics
   class QueryValidatorService
+    SCHEMA_FILE = Analytics::Engine.root / 'app' / 'services' / 'analytics' / 'query_schema.json.erb'
```
Some really valuable refactoring in this file!

jamesspeake commented on back/db/views/analytics_fact_project_statuses_v01.sql:
```diff
@@ -0,0 +1,41 @@
+         SELECT * FROM finished_statuses_for_continuous_projects
+     )
+
+SELECT *, timestamp::DATE as date
```
for consistency with the other fact tables we should use dimension_date_id as the date field

jamesspeake commented on back/db/views/analytics_fact_project_statuses_v01.sql:
```diff
@@ -0,0 +1,41 @@
+-- finished projects.
+
+WITH last_project_statuses AS -- project statuses that do not need to be calculated
+         (SELECT DISTINCT ON (item_id) item_id as project_id, action as status, acted_at as timestamp
```
For consistency with the other fact tables we should use dimension_project_id as the project column.

jamesspeake commented on back/engines/commercial/analytics/spec/models/analytics/fact_project_status_spec.rb:
```diff
@@ -0,0 +1,96 @@
+
+  let(:user) { create(:user) }
+
+  it { is_expected.to belong_to(:project) }
```
See previous comment about belonging to dimension_project

jamesspeake reviewed (changes requested):
This looks nice, some really good streamlining of the existing code in here too. However, it's just really the comments about consistency in dimension naming that I think needs addressing.

adessy committed:
CL-1881 Refactor to follow naming conventions

adessy committed:
Merge remote-tracking branch 'origin/CL-1881/fact-project-status' into CL-1881/fact-project-status

adessy committed:
Merge branch 'master' into CL-1881/fact-project-status

# Conflicts:
#	back/db/schema.rb
#	back/engines/commercial/analytics/app/services/analytics/query_schema.json.erb
#	back/engines/commercial/analytics/lib/query.rb

jamesspeake reviewed (approved):
All looks goo!

jamesspeake commented:
> All looks goo!

Good not goo!

adessy committed:
Merge branch 'master' into CL-1881/fact-project-status

# Conflicts:
#	back/db/schema.rb
#	back/engines/commercial/analytics/app/services/analytics/query_schema.json.erb
#	back/engines/commercial/analytics/lib/query.rb
